src/torrent.cpp

```diff
@@ -6797,10 +6806,13 @@ namespace libtorrent
 				}
 #endif
 				p->disconnect(errors::too_many_connections, op_bittorrent);
-				// we have to do this here because from the peer's point of
+				// we have to do this here because from the peer's point of view
 				// it wasn't really attached to the torrent, but we do need
 				// to let peer_list know we're removing it
 				remove_peer(p);
+				// the peer is already inserted and needs to be removed
+				TORRENT_ASSERT(sorted_find(m_connections, p) != m_connections.end());
+				m_connections.erase(sorted_find(m_connections, p));
```

Comment:
a few lines above L6748 you have `sorted_insert(m_connections, p);`. I thought that if this is a safe place to add to it (and invalidate the iterators) it should be safe to remove from it. This is important because at the end, the peer connection is not associated to the torrent, it's destroyed by the `session_impl::tick`, but when the times comes to call `torrent::disconnect_all`, some pointers in the loop are invalid